File: sdk/js/src/mesh.ts
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { mkdir, readFile, rename, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { ControlPlaneClient, ROLE_NODE, type Enrollment } from "./controlplane.ts";
import { credentialFromJSON, credentialToJSON, encodeAuthFrame, type MeshCredential } from "./credential.ts";
import { Identity } from "./identity.ts";

const IDENTITY_FILE = "identity.key";
const CREDENTIAL_FILE = "credential.json";

export interface AgentMeshOptions {
  /** Base URL of the control plane, e.g. https://hub.sam-mesh.dev. */
  controlPlaneUrl: string;
  /** Accept plaintext http:// to a non-loopback control plane. Off by default. */
  allowInsecure?: boolean;
  /**
   * Directory that keeps the identity key and the credential across
   * restarts. Without it the identity lives only in this process.
   */
  stateDir?: string;
  /** Use this identity instead of the persisted or a freshly generated one. */
  identity?: Identity;
  /** Role to enroll as. Defaults to ROLE_NODE. */
  role?: string;
  /** Operator-declared labels, e.g. { region: "eu" }. */
  labels?: Record<string, string>;
  /** Injection point for tests. */
  fetch?: typeof fetch;
}

export interface EnrollOptions extends AgentMeshOptions {
  /** A bootstrap token value, when the caller already holds it in memory. */
  bootstrapToken?: string;
  /** Path of a file holding the bootstrap token. Preferred over a value. */
  bootstrapTokenPath?: string;
  /** An OIDC ID token, for meshes that enroll identities interactively. */
  jwt?: string;
  /** Bounds the wait for an operator to approve a pending enrollment. */
  signal?: AbortSignal;
  /** Overrides the control plane's suggested poll interval while pending. */
  pollIntervalMs?: number;
}

/**
 * A member of the mesh: an identity, the credential the control plane minted
 * for it, and the client that keeps that credential fresh.
 *
 * This is the enrollment half of the SDK. Joining the mesh over libp2p and
 * serving or calling MCP tools build on top of it (see sdk/README.md).
 */
export class AgentMesh {
  readonly identity: Identity;
  readonly controlPlane: ControlPlaneClient;
  #credential: MeshCredential;
  readonly #stateDir: string | undefined;

  private constructor(identity: Identity, controlPlane: ControlPlaneClient, credential: MeshCredential, stateDir: string | undefined) {
    this.identity = identity;
    this.controlPlane = controlPlane;
    this.#credential = credential;
    this.#stateDir = stateDir;
  }

  get peerId(): string {
    return this.identity.peerId;
  }

  get credential(): MeshCredential {
    return this.#credential;
  }

  /**
   * Enrolls with the control plane and returns a member holding a credential.
   * Exactly one of bootstrapToken, bootstrapTokenPath or jwt must be given.
   */
  static async enroll(options: EnrollOptions): Promise<AgentMesh> {
    const given = [options.bootstrapToken, options.bootstrapTokenPath, options.jwt].filter((v) => v !== undefined).length;
    if (given !== 1) {
      throw new Error("exactly one of bootstrapToken, bootstrapTokenPath or jwt is required");
    }
    const identity = options.identity ?? (await loadIdentity(options.stateDir)) ?? Identity.generate();
    const controlPlane = newClient(options);
    const role = options.role ?? ROLE_NODE;

    let enrollment: Enrollment;
    if (options.jwt !== undefined) {
      enrollment = await controlPlane.register({ identity, jwt: options.jwt, role, ...labelsOf(options) });
    } else {
      const bootstrapToken = options.bootstrapTokenPath !== undefined ? (await readFile(options.bootstrapTokenPath, "utf8")).trim() : (options.bootstrapToken as string);
      enrollment = await controlPlane.enrollBootstrap({
        identity,
        bootstrapToken,
        role,
        ...labelsOf(options),
        ...(options.pollIntervalMs !== undefined ? { pollIntervalMs: options.pollIntervalMs } : {}),
        ...(options.signal !== undefined ? { signal: options.signal } : {}),
      });
    }

    // Widen trust from the one key the enrollment carries to every key the
    // control plane currently signs with, so peers holding credentials from
    // a retiring key still verify. Best effort, as in sam-node.
    let controlPlaneKeys = [enrollment.controlPlanePublicKey];
    try {
      controlPlaneKeys = await controlPlane.keys(controlPlaneKeys);
    } catch {
      // The enrollment key alone still works until the next sync.
    }

    const mesh = new AgentMesh(
      identity,
      controlPlane,
      {
        controlPlaneUrl: controlPlane.url.toString(),
        biscuit: enrollment.biscuit,
        expiration: enrollment.expiration,
        controlPlaneKeys,
        routerAddresses: enrollment.routerAddresses,
      },
      options.stateDir,
    );
    await mesh.save();
    return mesh;
  }

  /** Resumes a member from a state directory written by an earlier enroll(). */
  static async load(options: AgentMeshOptions & { stateDir: string }): Promise<AgentMesh> {
    const identity = options.identity ?? (await loadIdentity(options.stateDir));
    if (!identity) {
      throw new Error(`no identity in ${options.stateDir}; enroll first`);
    }
    const credential = credentialFromJSON(await readFile(join(options.stateDir, CREDENTIAL_FILE), "utf8"));
    return new AgentMesh(identity, newClient({ ...options, controlPlaneUrl: credential.controlPlaneUrl }), credential, options.stateDir);
  }

  /**
   * Trades the current biscuit for a fresh one and persists it. The control
   * plane redeems only the last biscuit it issued, so a lost refresh result
   * means re-enrolling; persisting before returning keeps that rare.
   */
  async refresh(): Promise<MeshCredential> {
    const result = await this.controlPlane.refresh({ identity: this.identity, biscuit: this.#credential.biscuit });
    let controlPlaneKeys = this.#credential.controlPlaneKeys;
    try {
      controlPlaneKeys = await this.controlPlane.keys(controlPlaneKeys);
    } catch {
      // Keep the previous set; a failed /keys sync must not cost the new biscuit.
    }
    this.#credential = { ...this.#credential, biscuit: result.biscuit, expiration: result.expiration, controlPlaneKeys };
    await this.save();
    return this.#credential;
  }

  /**
   * The frame that opens every stream to a peer: this member's biscuit plus
   * the service it wants (e.g. "mcp://calculator") and the agent it speaks for.
   */
  authFrame(targetService = "", agent = ""): Uint8Array {
    return encodeAuthFrame(this.#credential.biscuit, targetService, agent);
  }

  /** Writes identity and credential to the state directory, if one is configured. */
  async save(): Promise<void> {
    if (this.#stateDir === undefined) {
      return;
    }
    await mkdir(this.#stateDir, { recursive: true, mode: 0o700 });
    await writeAtomic(join(this.#stateDir, IDENTITY_FILE), this.identity.toLibp2pPrivateKey(), 0o600);
    await writeAtomic(join(this.#stateDir, CREDENTIAL_FILE), credentialToJSON(this.#credential), 0o600);
  }
}

function newClient(options: AgentMeshOptions): ControlPlaneClient {
  return new ControlPlaneClient({
    url: options.controlPlaneUrl,
    allowInsecure: options.allowInsecure ?? false,
    ...(options.fetch !== undefined ? { fetch: options.fetch } : {}),
  });
}

function labelsOf(options: AgentMeshOptions): { labels?: Record<string, string> } {
  return options.labels !== undefined ? { labels: options.labels } : {};
}

async function loadIdentity(stateDir: string | undefined): Promise<Identity | undefined> {
  if (stateDir === undefined) {
    return undefined;
  }
  try {
    return Identity.fromLibp2pPrivateKey(new Uint8Array(await readFile(join(stateDir, IDENTITY_FILE))));
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
      return undefined;
    }
    throw err;
  }
}

async function writeAtomic(path: string, data: Uint8Array | string, mode: number): Promise<void> {
  const tmp = `${path}.tmp`;
  await writeFile(tmp, data, { mode });
  await rename(tmp, path);
}

File: sdk/python/src/agent_mesh/mesh.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import os
import threading
import time
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import Mapping, Optional

from .controlplane import ROLE_NODE, ControlPlaneClient, Enrollment, Transport
from .credential import MeshCredential, encode_auth_frame
from .identity import Identity

_IDENTITY_FILE = "identity.key"
_CREDENTIAL_FILE = "credential.json"
# A saved credential with less validity left than this is not worth resuming; enroll again instead.
_REUSE_MIN_TTL_SECONDS = 5 * 60


@dataclass(frozen=True)
class ControlPlaneSync:
    """What one pull from the control plane changed."""

    # The trusted key set differs from before the pull.
    keys_changed: bool
    # The credential was re-issued because a rotation had happened since.
    refreshed: bool
    # The control plane's ban set, when /info answered.
    banned_peer_ids: Optional[list[str]]
    # When /info was asked (unix seconds); a ban recorded later cannot be in the answer.
    fetched_at: float
    # One entry per part that failed; empty when everything landed.
    errors: list[str] = field(default_factory=list)


class AgentMesh:
    """A member of the mesh: an identity, the credential the control plane
    minted for it, and the client that keeps that credential fresh. join()
    puts it on the mesh over libp2p."""

    def __init__(self, identity: Identity, control_plane: ControlPlaneClient, credential: MeshCredential, state_dir: Optional[Path]):
        self.identity = identity
        self.control_plane = control_plane
        self._credential = credential
        self._state_dir = state_dir

    @property
    def peer_id(self) -> str:
        return self.identity.peer_id

    @property
    def credential(self) -> MeshCredential:
        return self._credential

    @classmethod
    def enroll(
        cls,
        control_plane_url: str,
        *,
        bootstrap_token: Optional[str] = None,
        bootstrap_token_path: Optional[str | os.PathLike[str]] = None,
        jwt: Optional[str] = None,
        jwt_path: Optional[str | os.PathLike[str]] = None,
        state_dir: Optional[str | os.PathLike[str]] = None,
        identity: Optional[Identity] = None,
        role: str = ROLE_NODE,
        labels: Optional[Mapping[str, str]] = None,
        allow_insecure: bool = False,
        poll_interval: Optional[float] = None,
        cancel: Optional[threading.Event] = None,
        transport: Optional[Transport] = None,
    ) -> "AgentMesh":
        """Enrolls with the control plane and returns a member holding a credential.

        When state_dir already holds an unexpired credential from this control
        plane for the saved identity, that member is returned and no token is
        needed, so a program can call enroll on every start and read the token
        from its environment only on the first. Otherwise exactly one of
        bootstrap_token, bootstrap_token_path, jwt or jwt_path must be given; a
        token is better read from a file than passed as a value, and jwt_path
        also takes a platform's workload identity token, such as a Kubernetes
        projected service account token. Delete the state directory to enroll
        afresh, for instance with other labels."""
        state = Path(state_dir).expanduser() if state_dir is not None else None
        saved = _load_identity(state)
        identity = identity or saved or Identity.generate()
        control_plane = ControlPlaneClient(control_plane_url, allow_insecure=allow_insecure, transport=transport)
        if state is not None and saved is not None and saved.peer_id == identity.peer_id:
            credential = _load_credential(state)
            if credential is not None and credential.control_plane_url.rstrip("/") == control_plane.url.rstrip("/") and credential.time_to_live_seconds() > _REUSE_MIN_TTL_SECONDS:
                return cls(identity, control_plane, credential, state)
        given = sum(v is not None for v in (bootstrap_token, bootstrap_token_path, jwt, jwt_path))
        if given != 1:
            where = f" (no credential to resume in {state})" if state is not None else ""
            raise ValueError(f"exactly one of bootstrap_token, bootstrap_token_path, jwt or jwt_path is required{where}")

        enrollment: Enrollment
        if jwt is not None or jwt_path is not None:
            if jwt_path is not None:
                jwt = Path(jwt_path).expanduser().read_text(encoding="utf-8").strip()
            assert jwt is not None
            enrollment = control_plane.register(identity, jwt, role=role, labels=labels)
        else:
            if bootstrap_token_path is not None:
                bootstrap_token = Path(bootstrap_token_path).expanduser().read_text().strip()
            assert bootstrap_token is not None
            enrollment = control_plane.enroll_bootstrap(
                identity, bootstrap_token, role=role, labels=labels, poll_interval=poll_interval, cancel=cancel
            )

        # Widen trust from the one key the enrollment carries to every key the
        # control plane currently signs with, so peers holding credentials from
        # a retiring key still verify. Best effort, as in sam-node.
        control_plane_keys = [enrollment.control_plane_public_key]
        try:
            control_plane_keys = control_plane.keys(control_plane_keys)
        except Exception:  # noqa: BLE001 - the enrollment key alone still works until the next sync
            pass

        mesh = cls(
            identity,
            control_plane,
            MeshCredential(
                control_plane_url=control_plane.url,
                biscuit=enrollment.biscuit,
                expiration=enrollment.expiration,
                control_plane_keys=control_plane_keys,
                issued_under_keys=list(control_plane_keys),
                router_addresses=list(enrollment.router_addresses),
            ),
            state,
        )
        mesh.save()
        return mesh

    @classmethod
    def load(
        cls,
        state_dir: str | os.PathLike[str],
        *,
        identity: Optional[Identity] = None,
        allow_insecure: bool = False,
        transport: Optional[Transport] = None,
    ) -> "AgentMesh":
        """Resumes a member from a state directory written by an earlier enroll(),
        for a process that must never hold an enrollment token. The control plane
        URL comes from the saved credential."""
        state = Path(state_dir).expanduser()
        identity = identity or _load_identity(state)
        if identity is None:
            raise FileNotFoundError(f"no identity in {state}; enroll first")
        credential = _load_credential(state)
        if credential is None:
            raise FileNotFoundError(f"no credential in {state}; enroll first")
        control_plane = ControlPlaneClient(credential.control_plane_url, allow_insecure=allow_insecure, transport=transport)
        return cls(identity, control_plane, credential, state)

    def refresh(self) -> MeshCredential:
        """Trades the current biscuit for a fresh one and persists it. The control
        plane redeems only the last biscuit it issued, so a lost refresh result
        means re-enrolling; persisting before returning keeps that rare."""
        result = self.control_plane.refresh(self.identity, self._credential.biscuit)
        control_plane_keys = self._credential.control_plane_keys
        try:
            control_plane_keys = self.control_plane.keys(control_plane_keys)
        except Exception:  # noqa: BLE001 - a failed /keys sync must not cost the new biscuit
            pass
        self._credential = replace(
            self._credential,
            biscuit=result.biscuit,
            expiration=result.expiration,
            control_plane_keys=control_plane_keys,
            issued_under_keys=list(control_plane_keys),
        )
        self.save()
        return self._credential

    def add_trusted_key(self, key: bytes) -> bool:
        """Adopts a signing key announced by a KEY_ROTATION event, so peers whose
        credentials the new key signs verify before the next pull confirms it."""
        if any(bytes(k) == bytes(key) for k in self._credential.control_plane_keys):
            return False
        self._credential = replace(self._credential, control_plane_keys=[*self._credential.control_plane_keys, bytes(key)])
        return True

    def sync_control_plane(self) -> ControlPlaneSync:
        """The member's pull from the control plane, as sam-node's SyncControlPlane:
        the signing keys (verified against the set already trusted, so whoever
        answers the URL cannot become the trust root), a credential refresh when
        a rotation happened since it was issued, and /info for the router
        addresses and the ban set. Each part is attempted even when another
        fails; the errors are reported together. Gossip events only bring this
        forward; they are never the only way state arrives."""
        errors: list[str] = []
        keys_changed = False
        refreshed = False
        try:
            keys = self.control_plane.keys(self._credential.control_plane_keys)
            if not keys:
                raise ValueError("/keys returned no keys")
            keys_changed = {bytes(k) for k in keys} != {bytes(k) for k in self._credential.control_plane_keys}
            self._credential = replace(self._credential, control_plane_keys=list(keys))
            if self._credential.predates_rotation():
                self.refresh()
                refreshed = True
        except Exception as err:  # noqa: BLE001 - reported, the other parts still run
            errors.append(f"keys: {err}")

        banned_peer_ids: Optional[list[str]] = None
        # Taken before the request: a ban recorded after this instant cannot be
        # in the answer, so its absence must not be read as an unban.
        fetched_at = time.time()
        try:
            info = self.control_plane.info()
            if info.router_addresses:
                self._credential = replace(self._credential, router_addresses=list(info.router_addresses))
            banned_peer_ids = list(info.banned_peer_ids)
        except Exception as err:  # noqa: BLE001
            errors.append(f"info: {err}")
        if keys_changed or banned_peer_ids is not None:
            self.save()
        return ControlPlaneSync(keys_changed=keys_changed, refreshed=refreshed, banned_peer_ids=banned_peer_ids, fetched_at=fetched_at, errors=errors)

    def auth_frame(self, target_service: str = "", agent: str = "") -> bytes:
        """The frame that opens every stream to a peer: this member's biscuit plus
        the service it wants (e.g. "mcp://calculator") and the agent it speaks for."""
        return encode_auth_frame(self._credential.biscuit, target_service, agent)

    def join(self, **options):  # type: ignore[no-untyped-def]
        """Joins the mesh: connects to the routers in the credential, passes the
        auth handshake with them, reserves a relay slot and keeps the credential
        fresh. An async context manager to use under trio:

            async with mesh.join() as session: ...
        """
        from .session import join_mesh

        return join_mesh(self, **options)

    def save(self) -> None:
        """Writes identity and credential to the state directory, if one is configured."""
        if self._state_dir is None:
            return
        self._state_dir.mkdir(parents=True, exist_ok=True, mode=0o700)
        _write_atomic(self._state_dir / _IDENTITY_FILE, self.identity.to_libp2p_private_key())
        _write_atomic(self._state_dir / _CREDENTIAL_FILE, self._credential.to_json().encode())


def _load_identity(state: Optional[Path]) -> Optional[Identity]:
    if state is None:
        return None
    try:
        return Identity.from_libp2p_private_key((state / _IDENTITY_FILE).read_bytes())
    except FileNotFoundError:
        return None


def _load_credential(state: Path) -> Optional[MeshCredential]:
    try:
        return MeshCredential.from_json((state / _CREDENTIAL_FILE).read_text())
    except FileNotFoundError:
        return None


def _write_atomic(path: Path, data: bytes) -> None:
    tmp = path.with_name(path.name + ".tmp")
    fd = os.open(tmp, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
    with os.fdopen(fd, "wb") as f:
        f.write(data)
    os.replace(tmp, path)
